src/drivers/magnetometer/bosch/bmm350/BMM350.cpp

```diff
@@ -0,0 +1,683 @@
+	_state = STATE::RESET;
+	ScheduleClear();
+	ScheduleDelayed(1_ms);
+	return true;
+}
+
+void BMM350::print_status()
+{
+	I2CSPIDriverBase::print_status();
+
+	perf_print_counter(_reset_perf);
+	perf_print_counter(_bad_read_perf);
+	perf_print_counter(_self_test_failed_perf);
+}
+
+void BMM350::ParametersUpdate(bool force)
+{
+	if (_parameter_update_sub.updated() || force) {
+		parameter_update_s param_update;
+		_parameter_update_sub.copy(&param_update);
+		updateParams();
+		UpdateMagParams();
+	}
+}
+
+void BMM350::UpdateMagParams()
+{
+	uint8_t odr = GetODR(_param_bmm350_odr.get());
+	uint8_t avg = GetAVG(_param_bmm350_avg.get());
+
+	_mag_odr_mode = odr;
+	_mag_avg_mode = avg;
+	_mag_pad_drive = static_cast<uint8_t>(_param_bmm350_drive.get());
+	PX4_DEBUG("Set params odr = %d, avg = %d, drive = %d", _mag_odr_mode, _mag_avg_mode, _mag_pad_drive);
+}
+
+uint8_t BMM350::GetODR(int value)
+{
+	switch (value) {
+	case 0: return ODR_400HZ;
+
+	case 1: return ODR_200HZ;
+
+	case 2: return ODR_100HZ;
+
+	case 3: return ODR_50HZ;
+
+	case 4: return ODR_25HZ;
+
+	case 5: return ODR_12_5HZ;
+
+	case 6: return ODR_6_25HZ;
+
+	case 7: return ODR_3_125HZ;
+
+	case 8: return ODR_1_5625HZ;
+
+	default: return ODR_200HZ;
+	}
+}
+
+hrt_abstime BMM350::OdrToUs(uint8_t odr) {
+ switch (odr) {
+        case ODR_400HZ:
+            return 2500_us;
+        case ODR_200HZ:
+            return 5000_us;
+        case ODR_100HZ:
+            return 10000_us;
+        case ODR_50HZ:
+            return 20000_us;
+        case ODR_25HZ:
+            return 40000_us;
+        case ODR_12_5HZ:
+            return 80000_us;
+        case ODR_6_25HZ:
+            return 160000_us;
+        case ODR_3_125HZ:
+            return 320000_us;
+        case ODR_1_5625HZ:
+            return 640000_us;
+        default:
+            return 5000_us;
+    }
+}
+
+uint8_t BMM350::GetAVG(int value)
+{
+	switch (value) {
+	case 0: return AVG_NO_AVG;
+
+	case 1: return AVG_2;
+
+	case 2: return AVG_4;
+
+	case 3: return AVG_8;
+
+	default: return AVG_2;
+	}
+}
+
+
+int BMM350::probe()
+{
+	for (int i = 0; i < 3; i++) {
+		const uint8_t CMD = RegisterRead(Register::CMD);
+		const uint8_t CHIP_ID = RegisterRead(Register::CHIP_ID);
+
+		PX4_DEBUG("CMD: 0x%02hhX, CHIP_ID: 0x%02hhX", CMD, CHIP_ID);
+
+		if (CHIP_ID == chip_identification_number) {
+			PX4_DEBUG("Found chip");
+			return PX4_OK;
+
+		} else if (CHIP_ID == 0 && CMD == 0) {
+			PX4_DEBUG("Suspended, but found");
+			return PX4_OK;
+		}
+	}
+
+	return PX4_ERROR;
+}
+
+void BMM350::RunImpl()
+{
+	const hrt_abstime now = hrt_absolute_time();
+	ParametersUpdate();
+
+	switch (_state) {
+	case STATE::RESET: {
+			RegisterWrite(Register::CMD, SOFT_RESET);
+			_reset_timestamp = now;
+			_state = STATE::WAIT_FOR_RESET;
+			perf_count(_reset_perf);
+			ScheduleDelayed(10_ms);
+		}
+		break;
+
+	case STATE::WAIT_FOR_RESET: {
+			uint8_t chipId;
+
+			if ((chipId = RegisterRead(Register::CHIP_ID)) == chip_identification_number) {
```

Comment:
suggestion: could just call "probe()" here

Reply:
Hmm, regarding the values, it seems like I didn't check uncalibrated values.

nsh> listener vehicle_magnetometer

TOPIC: vehicle_magnetometer
 vehicle_magnetometer
    timestamp: 17710478 (0.109007 seconds ago)
    timestamp_sample: 17666390 (44088 us before timestamp)
    device_id: 15012873 (Type: 0xE5, I2C:1 (0x14))
    magnetometer_ga: [-0.22193, 0.03159, 0.24370]
    calibration_count: 3

nsh> listener sensor_mag

TOPIC: sensor_mag
 sensor_mag
    timestamp: 25610459 (0.003953 seconds ago)
    timestamp_sample: 25608889 (1570 us before timestamp)
    device_id: 15012873 (Type: 0xE5, I2C:1 (0x14))
    x: -2.21848
    y: -9.26652
    z: -12.72768
    temperature: 24.35856
    error_count: 0
    
 But i do see, that I forgot to turn off the self test for axis. Good catch, I will take a look, thanks!